Initial trace of a program with `torch.autograd.Function` is not printable

## 🐛 Bug

<!-- A clear and concise description of what the bug is. -->

If a trace has a BoundSymbol whose Symbol is registered to [AdHocExecutor](https://github.com/Lightning-AI/lightning-thunder/blob/f89cecac3b8a7e1e67913b1f0b53fa4c3b8322e8/thunder/extend/__init__.py#L276), then `print(thunder.last_traces(jitted)[0])` fails.

### To Reproduce

Steps to reproduce the behavior:

1. Add `print(thunder.last_traces(jitted)[0])` after https://github.com/Lightning-AI/lightning-thunder/blob/f89cecac3b8a7e1e67913b1f0b53fa4c3b8322e8/thunder/tests/test_jit_general.py#L1203
1. Run `pytest /path/to/test_jit_general.py -k "test_custom_autograd_function"`
2. See the error like the following one:

<!-- If you have a code sample, error messages, stack traces, please provide it here as well -->

```
>       print(thunder.last_traces(jitted)[0])

thunder/tests/test_jit_general.py:1204:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
thunder/core/trace.py:510: in __repr__
    return self.python(print_depth=-1)
thunder/core/trace.py:363: in python
    import_ctx, call_ctx, object_ctx = self._gather_ctxs()
thunder/core/trace.py:322: in _gather_ctxs
    bsym_import_ctx, bsym_call_ctx, bsym_object_ctx = bsym.gather_ctxs()
thunder/core/symbol.py:665: in gather_ctxs
    return self.import_ctx(), self._get_call_ctx(), self.object_ctx()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = t9 = MyLinear_139519205218000_0(x, weight, (2, 2))  # t9: "cpu f64[2, 2]"
  # t8 = ltorch.t(weight)  # t8: "cpu f64[2,...u f64[2, 2]"
  # t9 = ltorch.matmul(x, t8)  # t9: "cpu f64[2, 2]"
    # t9 = prims.matmul(x, t8)  # t9: "cpu f64[2, 2]"

    def import_ctx(self):
        # NOTE This initializes the context (Accessing these properties is a function call with the desired side effect)
        self._out_printables, self._arg_printables, self._kwarg_printables  # type: ignore
        if self.sym is not None and self.sym.python_impl is not None:
            # NOTE BoundSymbols of Symbols with a python_impl defined are run in Python, and are assumed
            #   to not need any imports to run properly, unless _import_prims is True
            if self.sym._print_as_impl:
                assert self.sym.module is not None  # TODO: Is this a valid assumption?
                module_name = self.sym.module.__name__
                import_ctx = {module_name: self.sym.module}
            else:
                import_ctx = {}
        elif self._call_ctx is not None:
            # NOTE If the call ctx was specified directly, then no import is needed to call the function
            import_ctx = {}
        else:
            # BoundSymbols of Symbols without Python implementations (either because they
            #   have Python implementations or defined call ctxs) are assumed to need
            #   a module import to run properly
>           assert self.sym.module is not None  # TODO: Is this a valid assumption?
E           AssertionError

thunder/core/symbol.py:625: AssertionError
=================================================================== short test summary info ====================================================================
FAILED thunder/tests/test_jit_general.py::test_custom_autograd_function - AssertionError
======================================================== 1 failed, 99 deselected, 29 warnings in 0.82s =========================================================
```

### Expected behavior

<!-- A clear and concise description of what you expected to happen. -->

### Environment

- PyTorch Version (e.g., 1.0): 2.6.0a0+git93aef68
- OS (e.g., Linux): Ubuntu 22.04
- How you installed PyTorch (`conda`, `pip`, source): source
- Build command you used (if compiling from source):
- Python version: 3.11
- CUDA/cuDNN version: 12.6 / 9.4.0
- GPU models and configuration: RTX 6000 Ada

### Additional context

<!-- Add any other context about the problem here. -->

--- a/thunder/core/symbol.py
+++ b/thunder/core/symbol.py
@@ -619,18 +619,22 @@ class BoundSymbol(BoundSymbolInterface):
             # NOTE If the call ctx was specified directly, then no import is needed to call the function
             import_ctx = {}
         else:
+            from thunder.extend import AdHocExecutor
             # BoundSymbols of Symbols without Python implementations (either because they
             #   have Python implementations or defined call ctxs) are assumed to need
             #   a module import to run properly
-            assert self.sym.module is not None  # TODO: Is this a valid assumption?
-            module_name = self.sym.module.__name__
-            import_ctx = {module_name: self.sym.module}
+            if isinstance(self.sym.executor, AdHocExecutor):
+                import_ctx = {}
+            else:
+                assert self.sym.module is not None  # TODO: Is this a valid assumption?
+                module_name = self.sym.module.__name__
+                import_ctx = {module_name: self.sym.module}
 
-            # TODO Include the other modules on the path?
-            # Also includes the root module of this (potential) submodule
-            if "." in module_name:
-                root_name = module_name.split(".")[0]
-                import_ctx[root_name] = sys.modules[root_name]
+                # TODO Include the other modules on the path?
+                # Also includes the root module of this (potential) submodule
+                if "." in module_name:
+                    root_name = module_name.split(".")[0]
+                    import_ctx[root_name] = sys.modules[root_name]
 
         self._import_ctx.update(import_ctx)
         return self._import_ctx

--- a/thunder/core/symbol.py
+++ b/thunder/core/symbol.py
@@ -620,6 +620,7 @@ class BoundSymbol(BoundSymbolInterface):
             import_ctx = {}
         else:
             from thunder.extend import AdHocExecutor
+
             # BoundSymbols of Symbols without Python implementations (either because they
             #   have Python implementations or defined call ctxs) are assumed to need
             #   a module import to run properly

